PR for second milstone: user interface

smcommits opened the pull request:
This PR is for the second milestone of Tic-tac-toe project. 
In this milestone, we worked on the feature branch name user-interface and we have created a dummy logic for the game

We have added the classes Board and Player. We have added helper methods to the respective classess, all the helper methods return _nil_ as we still haven't implemnted the game logic.

smcommits committed:
Initialize user inteface, Add User interface componenets

Intitialized the feature branch user interface
Added the componenets relating to user interface in the main.rb file
- Added a method to diplay instructions
- Added a mehtod to take input
- Added a method to display board

Commit:
Finalize dummy logic

Added class Board
- Added helper methods to board class
- Helper methods return nil
Added Player class
- Added switch_player method
- switch_player return nil
Initialize dummy logic in main.rb using dummy classes

smcommits committed:
Update dummy logic

Update dummy logic to depict the entire logic of the game
Returned to user_input to position
Added update_board method
Added ininite loop with a break clause of win? and tied? method

Commit:
Fix Linter Errors

Refactor explicit return statement to implicit
Refactor indentation

umairarshadbutt reviewed (changes requested):
# Status: REQUIRED CHANGES :warning: :construction: 

Hi @smcommits  and @Lameck1 
Good job with the changes! :tada: We are almost there.
We, however, do have a few changes to carry out before we proceed (see review comment).

##

Kindly work on the suggested changes 👇.
- [ ] Remove puts from other files because `main.rb` uses puts/print and get.chomp commands. ([screenshot](https://user-images.githubusercontent.com/29689273/98097182-7ae67b00-1eae-11eb-8cbd-3b52c13b4ce2.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))
- [ ] `draw` method is undefined. Please define it.
- [ ] Game tells which player turn it is (without computing who the winner is) 
([screenshot hint](https://user-images.githubusercontent.com/29689273/98098521-44a9fb00-1eb0-11eb-8fa2-6e8601422c46.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))
- [ ] Game asks the player to select from available moves([screenshot hint](https://user-images.githubusercontent.com/29689273/98098521-44a9fb00-1eb0-11eb-8fa2-6e8601422c46.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))
- [ ] Game displays board after the player move (without computing which spots on the board are taken)([screenshot hint](https://user-images.githubusercontent.com/29689273/98098521-44a9fb00-1eb0-11eb-8fa2-6e8601422c46.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))
- [ ] Game informs player if the selected move is a winning move (without computing the winning move)([screenshot hint](https://user-images.githubusercontent.com/29689273/98098521-44a9fb00-1eb0-11eb-8fa2-6e8601422c46.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))
- [ ] Game informs player if the selected move is a draw move (without computing draw)([screenshot hint](https://user-images.githubusercontent.com/29689273/98098521-44a9fb00-1eb0-11eb-8fa2-6e8601422c46.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))


##
- **When you have made the specified changes, please submit another code review. Thank you!**
##
**If you are stuck at any point or have any questions, let me know by leaving a comment here and mentioning me. I will be happy to help**
HAPPY CODING :computer:
**[Umair Arshad](https://umairarshad.me/)**
[Twitter](https://twitter.com/its_UmairArshad) | [LinkedIn](https://www.linkedin.com/in/umair-arshad-butt/) | [GitHub](https://github.com/umairarshadbutt)

smcommits commented:
@umairarshadbutt 
Thank you for the feedback!
I am afraid that I don't understand all the changes that have been requested. For instance, [the screenshot you provided](https://user-images.githubusercontent.com/29689273/98098521-44a9fb00-1eb0-11eb-8fa2-6e8601422c46.png) suggests that our game logic should not work, but all the changes that you requested require us to implement some of logic. 
Can you please elaborate on the changes requested?

umairarshadbutt commented:
### Hi @smcommits  and @Lameck1  


Thanks a lot for reaching me. I'm explaining the review with more explaination

##


- [ ] Remove puts from other files because `main.rb` uses puts/print and get.chomp commands. ([screenshot](https://user-images.githubusercontent.com/29689273/98097182-7ae67b00-1eae-11eb-8cbd-3b52c13b4ce2.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))
**Explanation:** You used the puts in `game_board.rb`. Remove the puts from `game_board.rb`. I attached the screenshot which is explaining that you can use puts and print only in main.rb for this project
- [ ] `draw` method is undefined. Please define it.
**Explanation:** You have not defined the draw method. When I run the project on my local machine you can see the issue on the attached [image](https://user-images.githubusercontent.com/29689273/98102280-488c4c00-1eb5-11eb-97b7-20e4081c5379.png)

- [ ] Game tells which player turn it is (without computing who the winner is) 
([screenshot hint](https://user-images.githubusercontent.com/29689273/98098521-44a9fb00-1eb0-11eb-8fa2-6e8601422c46.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))
**Explanation:** When we are playing the game there is no instruction on who is playing. Please tell the players about their moves.
- [ ] Game asks the player to select from available moves([screenshot hint](https://user-images.githubusercontent.com/29689273/98098521-44a9fb00-1eb0-11eb-8fa2-6e8601422c46.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))
**Explanation:** Player should be informed about available moves by displaying the board. But in your project board is not available in console.
- [ ] Game displays board after the player move (without computing which spots on the board are taken)([screenshot hint](https://user-images.githubusercontent.com/29689273/98098521-44a9fb00-1eb0-11eb-8fa2-6e8601422c46.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))
**Explanation:**  Board is missing in console
- [ ] Game informs player if the selected move is a winning move (without computing the winning move)([screenshot hint](https://user-images.githubusercontent.com/29689273/98098521-44a9fb00-1eb0-11eb-8fa2-6e8601422c46.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))
**Explanation:**  There is no winning message.  Winning logic will be defined in the next milestone but you should decide where and how you will use it with the message.
- [ ] Game informs player if the selected move is a draw move (without computing draw)([screenshot hint](https://user-images.githubusercontent.com/29689273/98098521-44a9fb00-1eb0-11eb-8fa2-6e8601422c46.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))
**Explanation:**  There is no draw message. Draw logic will be defined in the next milestone but you should decide where and how you will use it with the message.




##
- **When you have made the specified changes, please submit another code review. Thank you!**
##
**If you are stuck at any point or have any questions, let me know by leaving a comment here and mentioning me. I will be happy to help**
HAPPY CODING :computer:
**[Umair Arshad](https://umairarshad.me/)**
[Twitter](https://twitter.com/its_UmairArshad) | [LinkedIn](https://www.linkedin.com/in/umair-arshad-butt/) | [GitHub](https://github.com/umairarshadbutt)

smcommits commented:
Hey @umairarshadbutt 

The game informs the player if the selected move is a winning move (without computing the winning move)(screenshot hint from Pathwright)
Explanation: There is no winning message. Winning logic will be defined in the next milestone but you should decide where and how you will use it with the message.

The game informs the player if the selected move is a draw move (without computing draw)(screenshot hint from Pathwright)
Explanation: There is no draw message. Draw logic will be defined in the next milestone but you should decide where and how you will use it with the message.

I am sorry, I find the above two conditions contradictory because a game can either return a win or a tie. Also, for returning a winning or a tie message, we need to compute the positons. Can you please elaborate on the issue?

umairarshadbutt commented:
##
Hi @smcommits 
You are right the game will return a value for winning or draw. **But the suggested requirement is not about the logic, it is about the user interface.** 

As I mentioned in the First review **(without computing the winning move)** and **(without computing the draw**. It means you need to update or design a message for the required suggestion.  
##
**If you are stuck at any point or have any questions, let me know by leaving a comment here and mentioning me. I will be happy to help**
HAPPY CODING :computer:
**[Umair Arshad](https://umairarshad.me/)**
[Twitter](https://twitter.com/its_UmairArshad) | [LinkedIn](https://www.linkedin.com/in/umair-arshad-butt/) | [GitHub](https://github.com/umairarshadbutt)

smcommits committed:
Add win method, update method, get_name method, get_user_input method

Added a win method to decide if someone has won
Added an update method to update the board
Added the get_name method to get the name of the users
Added the get_user_input method to get the psoitonal inputs from the user

Commit:
Fix Linter Errors, Refactor get_name method

Fixed indentation and removed trailing whitespaces
Changed method name to user_name

smcommits commented:
@umairarshadbutt Okay! That put things in perspective. Thank you for the explanation. We will implement the changes and request a reveiw.

arslanbisharat reviewed (changes requested):
# Request changes 😳

I am very happy to see this project and I can see you worked hard on this project but I am sorry to inform you that I cannot approve this project because it is not completed yet. You need to work more on it. 
_It's okay. We are here to learn. Don't get discouraged. Work even harder. You are almost there.😄_

**Please have a look at the suggestions to get more understanding.**

- [ ] Remove puts from other files because `main.rb` uses **puts/print** and **get.chomp** commands. 
![1](https://user-images.githubusercontent.com/58504655/98127103-21934180-1ed8-11eb-8c2f-015260fa4feb.png)

- [ ] Game asks the player to select from available moves

![2](https://user-images.githubusercontent.com/58504655/98127177-37086b80-1ed8-11eb-9a64-707f545ca788.png)


**Please make the changes and request re-review for the given project**
Stay safe and give me a smile 😇

### Happy Coding
---------------------------------------------------------------------------------------------------------------------------------
**Regards,
Arslan Bisharat
slack: arslanbisharat**

arslanbisharat commented on bin/main.rb:
```diff
@@ -1,2 +1,45 @@
+require_relative '../lib/game_board'
+
+def display_instructions
+  puts 'This is a tic-tac-toe game'
+  puts 'Press the Enter key to start the game'
+  puts 'Player 1 is "X" and Player 2 is "O"'
+  puts
```
- [ ] Remove puts

smcommits replied:
Those puts are in the main file, Can you please explain why should I remove them?

arslanbisharat commented on bin/main.rb:
```diff
@@ -1,2 +1,45 @@
+end
+
+def user_name
+  print 'Enter the name of first player: '
+  name1 = gets.chomp
+  print 'Enter the name of second player: '
+  name2 = gets.chomp
+  puts "#{name1} will be using 'X' and #{name2} will be using 'O'"
+  [name1, name2]
```
- [ ] Remove puts and print. This is a requirement from the pathway.

arslanbisharat commented on bin/main.rb:
```diff
@@ -1,2 +1,45 @@
+
+def get_user_input(turn, name1, name2)
+  if turn == 'player_a'
+    print "#{name1}, please enter your position: "
```
- [ ] Remove puts and print. This is a requirement from the pathway.

arslanbisharat commented on bin/main.rb:
```diff
@@ -1,2 +1,45 @@
+  if turn == 'player_a'
+    print "#{name1}, please enter your position: "
+  else
+    print "#{name2}, please enter your position: "
```
- [ ] Remove puts and print. This is a requirement from the pathway.

arslanbisharat commented on bin/main.rb:
```diff
@@ -1,2 +1,45 @@
+
+  count += 1
+end
+puts("#{name1} is a winner")
+puts('It is a tie')
```
- [ ] Remove puts and print. This is a requirement from the pathway.

arslanbisharat commented on lib/game_board.rb:
```diff
@@ -0,0 +1,35 @@
+  end
+
+  def draw
+    puts "|#{board[0]} | #{board[1]} | #{board[2]}|"
+    puts '-----------'
+    puts "|#{board[3]} | #{board[4]} | #{board[5]}|"
+    puts '-----------'
+    puts "|#{board[6]} | #{board[7]} | #{board[8]}|"
```
- [ ] Remove puts and print. This is a requirement from the pathway.

arslanbisharat commented on bin/main.rb:
```diff
@@ -1,2 +1,45 @@
+
+def user_name
+  print 'Enter the name of first player: '
+  name1 = gets.chomp
```
- [ ] Remove gets.chomp. This is a requirement from the pathway.

arslanbisharat commented on bin/main.rb:
```diff
@@ -1,2 +1,45 @@
+  print 'Enter the name of first player: '
+  name1 = gets.chomp
+  print 'Enter the name of second player: '
+  name2 = gets.chomp
```
- [ ] Remove gets.chomp. This is a requirement from the pathway.

arslanbisharat commented on bin/main.rb:
```diff
@@ -1,2 +1,45 @@
+  else
+    print "#{name2}, please enter your position: "
+  end
+  position = gets.chomp
```
- [ ] Remove gets.chomp. This is a requirement from the pathway.

smcommits commented:
Hey! @arslanbisharat  thank you for the review. 

 **Game asks the player to select from available moves**
Can you please elaborate on the above-suggested change as the game is already asking the player to input based on the available moves? 
![This part of the code validates and returns the available inputs](https://user-images.githubusercontent.com/59168430/98128327-ccf2c500-1edd-11eb-9035-ce0e1f54e96b.png)
This part of the code validates and returns the available inputs

![2020-11-04-203808_780x45_scrot-framed](https://user-images.githubusercontent.com/59168430/98128422-ed228400-1edd-11eb-9cb8-03a3809413f1.png)
Microverse Pathwright suggests to us that we can use gets.chomp and puts only in main.rb file. Why have you asked us to remove them? Can you please elaborate on those?

arslanbisharat commented on bin/main.rb:
```diff
@@ -1,2 +1,45 @@
+  puts 'This is a tic-tac-toe game'
+  puts 'Press the Enter key to start the game'
+  puts 'Player 1 is "X" and Player 2 is "O"'
+  puts
```
I am sorry. Please ignore this. You should remove `puts/print` from **game_board.rb** only.

arslanbisharat commented:
You are right that there will be a method(**valid_move)**, which will validate the input according to the board. But the suggested requirement is not about the logic, it is about the user interface.
As previous TSEs mentioned in the First review and Re-review(Game asks the player to select from available moves). It means you need to update or design a message for the required suggestion.

smcommits commented:
@arslanbisharat So do you want me to display available moves and ask the user to select from it. Also, can you please elaborate on the request to remove chomp and puts?

arslanbisharat commented:
Remove puts from other files because main.rb uses puts/print and get.chomp commands. ([screenshot ](https://user-images.githubusercontent.com/29689273/98097182-7ae67b00-1eae-11eb-8cbd-3b52c13b4ce2.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))
Explanation: You used the puts in game_board.rb. Remove the puts from game_board.rb. I attached the screenshot which is explaining that you can use puts and print only in main.rb for this project

![image](https://user-images.githubusercontent.com/58504655/98136274-923f5b80-1ee2-11eb-813a-0f70ee4802ca.png)

smcommits committed:
Add dummy methods, update interface in main.rb

Added dummy methods
 - position_taken?
 - valid_move?
 - update_board

 Added interface to main.rb
 - Added input validation
 - Moved draw method from board library to main.rb

layzrd reviewed (approved):
### Status: Approved 🙌 🎇
Hi @Lameck1  and @smcommits , Congratulations 🎊 👏, you did very well on this project. ✨ Clap 👏👏👏 for accepting players' names.

Now, you can move ahead.
Happy coding 👨‍💻

[Abdulaziz Redi Ali](https://azizali.ml/)
